llmstudio/engine/providers/azure.py

```diff
@@ -44,48 +56,284 @@ async def generate_client(
                     api_version=request.api_version or self.API_VERSION,
                 )
-                return await asyncio.to_thread(
-                    client.chat.completions.create,
-                    model=request.model,
-                    messages=(
-                        [{"role": "user", "content": request.chat_input}]
-                        if isinstance(request.chat_input, str)
-                        else request.chat_input
-                    ),
-                    functions=request.functions,
-                    function_call="auto" if request.functions else None,
-                    stream=True,
-                    response_format=request.response_format,
-                    **request.parameters.model_dump(),
-                )
+
+            messages = self.prepare_messages(request)
+
+            return await asyncio.to_thread(
+                client.chat.completions.create,
+                model=request.model,
+                messages=messages,
+                stream=True,
+                **(
+                    {}
+                    if self.is_llama
+                    else {
+                        "tools": request.tools,
+                        "tool_choice": "auto" if request.tools else None,
+                        "response_format": request.response_format,
+                    }
+                ),
+                **request.parameters.model_dump(),
+            )
 
         except openai._exceptions.APIError as e:
             raise HTTPException(status_code=e.status_code, detail=e.response.json())
 
+    def prepare_messages(self, request: AzureRequest):
+        if self.is_llama:
+            user_message = self.convert_to_openai_format(request.chat_input)
+            content = "<|begin_of_text|>"
+            content = self.add_system_message(user_message, content, request.tools)
+            content = self.add_conversation(user_message, content)
+            return [{"role": "user", "content": content}]
+        else:
+            return (
+                [{"role": "user", "content": request.chat_input}]
+                if isinstance(request.chat_input, str)
+                else request.chat_input
+            )
+
     async def parse_response(
         self, response: AsyncGenerator, **kwargs
     ) -> AsyncGenerator[str, None]:
+        if self.is_llama and self.has_tools:
+            async for chunk in self.handle_tool_response(response, **kwargs):
+                yield chunk
+        else:
+            for chunk in response:
+                yield chunk.model_dump()
+
+    async def handle_tool_response(
+        self, response: AsyncGenerator, **kwargs
+    ) -> AsyncGenerator[str, None]:
+        saving = False
+        function_name = ""
+        function_response = False
         for chunk in response:
-            yield chunk.model_dump()
+            content = chunk.choices[0].delta.content
+            if content is not None:
+                if "§" in content:
```

Comment:
shouldn't it be a regex? what if the function has the "section sign" in it's call?

`# Example dataset
data = {
    'feature1': np.random.rand(100),
    'feature2': np.random.rand(100),
    'feature3': np.random.rand(100),
    '§': np.random.rand(100)  # Target variable with section sign
}
`